File: backend/usecase/interactor/discounts_interactor.go
package interactor

import (
	"github.com/Cirivas/desafio-walmart/models"
	"github.com/Cirivas/desafio-walmart/usecase/presenter"
	"github.com/Cirivas/desafio-walmart/usecase/repository"
)

type discountsInteractor struct {
	DiscountsRepository repository.DiscountsRepository
	DiscountsPresenter  presenter.DiscountsPresenter
}

type DiscountsInteractor interface {
	Get(brand string) (*models.Discount, error)
}

func NewDiscountsInteractor(r repository.DiscountsRepository, p presenter.DiscountsPresenter) DiscountsInteractor {
	return &discountsInteractor{r, p}
}

func (di *discountsInteractor) Get(brand string) (*models.Discount, error) {
	discount, err := di.DiscountsRepository.Find(brand)

	if err != nil {
		return nil, err
	}

	return di.DiscountsPresenter.ResponseDiscount(discount), nil
}

File: frontend/src/usecase/interactor/discounts.interactor.ts
import { Discount } from "../../domain/discount";
import { DiscountsPresenter } from "../presenter/discounts.presenter";
import { DiscountsRepository } from "../repository/discounts.repository";

export interface DiscountsInteractor {
  getFromBrand(brand: string): Promise<Discount>;
}
export function useDiscountsInteractor(
  dr: DiscountsRepository,
  dp: DiscountsPresenter
): DiscountsInteractor {
  return {
    getFromBrand: (brand: string) =>
      dr.find(brand).then((result) => dp.responseDiscount(result)),
  };
}
